erpnext/selling/doctype/sales_order/sales_order.py

```diff
@@ -1413,4 +1413,66 @@ def update_order_status():
 
 	for order in orders:
 		order_doc = frappe.get_doc("Sales Order", order)
-		order_doc.run_method("check_overdue_status")
\ No newline at end of file
+		order_doc.run_method("check_overdue_status")
+
+def validate_delivery_window(doc, method):
+	from erpnext.stock.doctype.delivery_trip.delivery_trip import get_delivery_window
+	if not frappe.db.get_single_value("Delivery Settings", "send_delivery_window_warning"):
+		return
+
+	if not (doc.get("delivery_start_time") and doc.get("delivery_end_time")):
+		return
+
+	if not doc.get("customer"):
+		return
+
+	delivery_window = get_delivery_window(customer=doc.get("customer"))
+	delivery_start_time = delivery_window.delivery_start_time
+	delivery_end_time = delivery_window.delivery_end_time
+
+	if not (delivery_start_time and delivery_end_time):
+		return
+
+	if to_timedelta(doc.delivery_start_time) < to_timedelta(delivery_start_time) \
+		or to_timedelta(doc.delivery_end_time) > to_timedelta(delivery_end_time):
+		if method == "validate":
+			frappe.msgprint(_("The delivery window is set outside the customer's default timings"))
+		elif method == "on_submit":
+			# send an email notifying users that the document is outside the customer's delivery window
+			role_profiles = ["Fulfillment Manager"]
+			role_profile_users = frappe.get_all("User", filters={"role_profile_name": ["IN", role_profiles]}, fields=["email"])
+			role_profile_users = [user.email for user in role_profile_users]
+
+			accounts_managers = get_users_with_role("Accounts Manager")
+			system_managers = get_users_with_role("System Manager")
+
+			recipients = list(set(role_profile_users + accounts_managers) - set(system_managers))
+
+			if not recipients:
+				return
+
+			# form the email
+			subject = _("[Info] An order may be delivered outside a customer's preferred delivery window")
+			message = _("""An order ({name}) has the following delivery window: {doc_start} - {doc_end}<br><br>
+				While the default delivery window is {customer_start} - {customer_end}""".format(
+					name=frappe.utils.get_link_to_form(doc.doctype, doc.name),
+					doc_start=get_time(doc.delivery_start_time).strftime("%I:%M %p"),
+					doc_end=get_time(doc.delivery_end_time).strftime("%I:%M %p"),
+					customer_start=get_time(delivery_start_time).strftime("%I:%M %p"),
+					customer_end=get_time(delivery_end_time).strftime("%I:%M %p"),
+				))
+
+			frappe.sendmail(recipients=recipients, subject=subject, message=message)
+
+def get_users_with_role(role):
+	# returns users with the specified role
+	user_list = frappe.get_all("User", fields=["`tabUser`.name"],
+		filters = [
+			["Has Role", "role", "=", role],
+			["User", "name", "not in", ["Guest", "Administrator"]],
+			["User", "enabled", "=", 1]
+		],
+		as_list=1
+	)
+
+	return [user for users in user_list for user in users]
```

Comment:
Move this to  `erpnext/erpnext/utils.py` since these are general functions used in multiple doctypes